File: C#/Dapper_ORM/DAL/Repositories/ProductRepository.cs
using DAL.Entities;
using Dapper;
using System.Data.SqlClient;


namespace DAL.Repositories;

public class ProductRepository /*: GenericRepository<Product, int>*/
{
    private readonly string _connectionString;
    public ProductRepository(string connectionString) /*: base(connectionString, "Products", "ProductID")*/
    {
        _connectionString = connectionString;
    }

    [Obsolete]
    public async Task<List<Product>> GetAllAsync()
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"SELECT * FROM Products";
            var result = await connection.QueryAsync<Product>(query);
            return result.ToList();
        }
    }
    [Obsolete]
    public async Task<Product?> GetByIdAsync(int id)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            connection.Open();
            var query = $"SELECT * FROM Products WHERE ProductID = @Id";
            return await connection.QueryFirstOrDefaultAsync<Product>(query, new { Id = id });
        }
    }
    [Obsolete]
    public async Task<Product> AddAsync(Product entity)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"INSERT INTO Products (ProductName, SupplierID, CategoryID, QuantityPerUnit, UnitPrice, UnitsInStock, UnitsOnOrder, ReorderLevel, Discontinued)" +
                $" VALUES (@ProductName, @SupplierID, @CategoryID, @QuantityPerUnit, @UnitPrice, @UnitsInStock, @UnitsOnOrder, @ReorderLevel, @Discontinued); SELECT CAST(SCOPE_IDENTITY() as int);";
            var id = await connection.QuerySingleAsync<int>(query, entity);
            var propertyInfo = typeof(Product).GetProperty("ProductID");
            propertyInfo.SetValue(entity, id);
            return entity;
        }
    }
    [Obsolete]
    public async Task UpdateAsync(Product entity)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"UPDATE Products SET ProductName = @ProductName, SupplierID = @SupplierID, CategoryID = @CategoryID, QuantityPerUnit = @QuantityPerUnit, UnitPrice = @UnitPrice," +
                $" UnitsInStock = @UnitsInStock, UnitsOnOrder = @UnitsOnOrder, ReorderLevel = @ReorderLevel, Discontinued = @Discontinued WHERE ProductID =  @ProductID";
            await connection.ExecuteAsync(query, entity);
        }
    }
    [Obsolete]
    public async Task DeleteAsync(int id)
    {
        using (var connection = new SqlConnection(_connectionString))
        {
            await connection.OpenAsync();
            var query = $"DELETE FROM Products WHERE ProductID = @Id";
            await connection.ExecuteAsync(query, new { Id = id });
        }
    }
}

File: Go/GORM_ORM/dal/repositories/productRepository.go
package repositories

import (
	"GORM_ORM/dal/entities"

	"gorm.io/gorm"
)

type ProductRepository interface {
	GetAll() ([]entities.Product, error)
	GetByID(id int) (*entities.Product, error)
	Create(entity *entities.Product) error
	Update(entity *entities.Product) error
	Delete(id int) error
}

type productRepository struct {
	genericRepo GenericRepository[entities.Product]
}

func NewProductRepository(db *gorm.DB) ProductRepository {
	return &productRepository{
		genericRepo: NewGenericRepository[entities.Product](db),
	}
}

func (r *productRepository) GetAll() ([]entities.Product, error) {
	return r.genericRepo.GetAll()
}

func (r *productRepository) GetByID(id int) (*entities.Product, error) {
	return r.genericRepo.GetByID(id)
}

func (r *productRepository) Create(entity *entities.Product) error {
	return r.genericRepo.Create(entity)
}

func (r *productRepository) Update(entity *entities.Product) error {
	return r.genericRepo.Update(entity)
}

func (r *productRepository) Delete(id int) error {
	return r.genericRepo.Delete(id)
}
